### ISSUE TITLE
Some things we can learn from pytypes

### ISSUE THREAD

**Issue by gvanrossum** [dropbox/pyannotate#38]
I checked the profiling hook in [pytypes](https://github.com/Stewori/pytypes/blob/master/pytypes/type_util.py#L1893), and they do two things we should also do:
- save and restore the previously active profiler
- in a thread, when the profiling callback is run after profiling has been stopped, reset the hook
  (note: this should only be done after it's been stopped -- not after it's been paused)

**Comment by gvanrossum** [comment]
Moreover if a previous profiler is already running they pass events to that too!

**Comment by gvanrossum** [comment]
Lots of other stuff too, e.g. their utilities for finding the actual function. (OTOH I don't see any downsampling or other speed hacks in their code.)

**Comment by Stewori** [comment]
Yes, the downsampling idea is something we can learn from pyannotate. I will surely add this.
